### PR TITLE
ADR-0019: GPIO

### PR DESCRIPTION
**Pull request by frenck**
ADR to enforce disallowing integrations that rely on GPIO to integrate into Home Assistant.

Based on the discussion in: #596

### CODE REVIEW

File: adr/0019-GPIO.md [hunk 1]
```diff
@@ -0,0 +1,27 @@
+# 0018. GPIO
```
Comment (frenck) [inline adr/0019-GPIO.md:1]:
```suggestion
# 0019. GPIO
```

File: adr/0019-GPIO.md [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+  but it adds the same amount of maintenance and review time to the core.
+- Most of these integrations are unmaintained.
+- The use of GPIO via containers (including our Home Assistant OS, Container
+  & Supervised installation methods) is often unstable or complicated.
+- Support for GPIO on specific platforms/SBCs: milage varies (e.g., Raspberry Pi on 64 bits platforms).
+- The use of these sensors, in general, is better supported and done via dedicated projects like ESPHome.
+
+## Decision
+
+We no longer accepting integrations that integrate with devices over GPIO.
```
Comment (MartinHjelmare) [inline adr/0019-GPIO.md:23]:
```suggestion
We no longer accept integrations that integrate with devices over GPIO.
```

**Review by pvizeli — APPROVED** [review]

**Comment by spacegaier** [comment]
Should something be mentioned how we want to proceed with the existing ones? My understanding is they will be left for now, unless they are abandoned (in which case there will be a deprecation period I assume to give the few users a change to look at ESPHome?).

**Comment by frenck** [comment]
> Should something be mentioned how we want to proceed with the existing ones? 

Yeah, we just chatted about that a bit... I'm updating this PR in a minute...

File: adr/0019-GPIO.md [hunk 3]
```diff
@@ -20,8 +20,12 @@ We currently have a lot of integrations that use or are based on GPIO (including
...
-We no longer accepting integrations that integrate with devices over GPIO.
+- We no longer accept integrations that integrate with devices over GPIO.
+- We identify, deprecate and remove integrations that use GPIO.
+- It will still be possible to have custom integrations that use GPIO.
 
 ## Consequences
 
-No longer accepting integrations that integrate with devices over GPIO.
+We will no longer accept integrations that integrate with devices over GPIO as
+a core integration, those will have to be maintained as cutom integrations,
```
Comment (MartinHjelmare) [inline adr/0019-GPIO.md:30]:
```suggestion
a core integration, those will have to be maintained as custom integrations,
```

**Review by MartinHjelmare — APPROVED** [review]

**Review by zsarnett — APPROVED** [review]

File: adr/0019-GPIO.md [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+## Context
+
+We currently have a lot of integrations that use or are based on GPIO (including SPI and i2c busses).
+
+- All of these integrations have a low usage count (source Home Assistant Analytics),
+  but it adds the same amount of maintenance and review time to the core.
+- Most of these integrations are unmaintained.
+- The use of GPIO via containers (including our Home Assistant OS, Container
+  & Supervised installation methods) is often unstable or complicated.
+- Support for GPIO on specific platforms/SBCs: milage varies (e.g., Raspberry Pi on 64 bits platforms).
```
Comment (raman325) [inline adr/0019-GPIO.md:18]:
```suggestion
- Support for GPIO on specific platforms/SBCs: mileage varies (e.g., Raspberry Pi on 64 bits platforms).
```

**Review by raman325 — APPROVED** [review]

File: adr/0019-GPIO.md (hunk 4 above)
Comment (frenck) [reply adr/0019-GPIO.md]: Thanks! 👍

**Review by balloob — APPROVED** [review]

File: adr/0019-GPIO.md [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+## Status
+
+Accepted
+
+## Context
+
+We currently have a lot of integrations that use or are based on GPIO (including SPI and i2c busses).
+
+- All of these integrations have a low usage count (source Home Assistant Analytics),
+  but it adds the same amount of maintenance and review time to the core.
```
Comment (TomBrien) [inline adr/0019-GPIO.md:14]:
```suggestion
  but add the same amount of maintenance and review time to the core.
```

**Review by bdraco — APPROVED** [review]

File: adr/0019-GPIO.md [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+  & Supervised installation methods) is often unstable or complicated.
+- Support for GPIO on specific platforms/SBCs: mileage varies (e.g., Raspberry Pi on 64 bits platforms).
+- The use of these sensors, in general, is better supported and done via dedicated projects like ESPHome.
+
+## Decision
+
+- We no longer accept integrations that integrate with devices over GPIO.
+- We identify, deprecate and remove integrations that use GPIO.
+- It will still be possible to have custom integrations that use GPIO.
+
```
Comment (TomBrien) [inline adr/0019-GPIO.md:26]: Does this need a comment on whether HAOS will continue to provide and support a means of accessing GPIO in the long term

(Not sure if this applies but I assume there is some effort in HAOS to provide reliable pass through of those pins)

**Review by TomBrien — APPROVED** [review]
Seems like totally common sense as not all hardware has GPIO and ESPhome is a thing.

File: adr/0019-GPIO.md (hunk 5 above)
Comment (frenck) [reply adr/0019-GPIO.md:14]: `it adds` or `adds`, but just `add` is incorrect in this case :)

File: adr/0019-GPIO.md (hunk 6 above)
Comment (frenck) [reply adr/0019-GPIO.md:26]: I dunno, I've kept that out of scope for this ADR (which is focused solely on the integrations parts). I don't see a reason to remove support from that 🤷

File: adr/0019-GPIO.md (hunk 5 above)
Comment (TomBrien) [reply adr/0019-GPIO.md:14]: Take out the part before the comma and you have "All of these integrations ... add the same amount of ..."

I kinda see the "it" but to me it's confusing as the sentence started with a plural subject but seemed to flipped to singular. Alternatively, something like "...to support these integrations"  on the end of the sentence works and looking at it now I think that's the point?

Comment (frenck) [reply adr/0019-GPIO.md:14]: You are right ☺️ 
Thanks!

File: adr/0019-GPIO.md [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+GPIO, in this case, means interfacing directly with the individual GPIO pins
+(or I/O lines) on the board running Home Assistant. This includes bus
+communications like I2C and SPI.
+
+Devices connected on GPIO that are exposed on the system as regular serial
+devices, are exempted from this ADR.
```
Comment (bachya) [inline adr/0019-GPIO.md:32]:
```suggestion
devices are exempted from this ADR.
```

File: adr/0019-GPIO.md [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+
+## Consequences
+
+We will no longer accept integrations that integrate with devices over GPIO as
+a core integration, those will have to be maintained as custom integrations,
```
Comment (bachya) [inline adr/0019-GPIO.md:37]:
```suggestion
a core integration; those will have to be maintained as custom integrations,
```

**Review by bachya — APPROVED** [review]
Small grammatical tweaks.

**Pull request merged**